### PR TITLE
Async client

### PR DESCRIPTION
**Pull request by eliseygusev**

### CODE REVIEW

File: .pylintrc [hunk 1]
```diff
@@ -1,2 +1,2 @@
 [MASTER]
-disable=fixme,logging-fstring-interpolation
\ No newline at end of file
+disable=fixme,logging-fstring-interpolation,duplicate-code
```
Comment (eliseygusev) [inline .pylintrc:2]: added "duplicate-code" here. might be dangerous, but without it pylint flags all the code (since the functionality is basically copy-pasted). but I don't think it it necessary to abstract functionality even further.

File: dune_client/base_client.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import logging.config
+from json import JSONDecodeError
+from typing import Any, Union
+
+from aiohttp import ClientResponse
+from requests import Response
+
+
+# pylint: disable=too-few-public-methods
+class BaseDuneClient:
```
Comment (eliseygusev) [inline dune_client/base_client.py:17]: added base client which basically has logger, some constants and common function for unwrapping response

File: dune_client/interface.py [hunk 3]
```diff
@@ -18,3 +18,16 @@ def refresh(self, query: Query) -> ResultsResponse:
         Executes a Dune query, waits till query execution completes,
         fetches and returns the results.
         """
+
+
+# pylint: disable=too-few-public-methods
+class AsyncDuneInterface(ABC):
```
Comment (eliseygusev) [inline dune_client/interface.py:24]: same as sync interface but has async interface for refresh?
also, may be it makes sense to add more functions to that interface?

File: dune_client/base_client.py (hunk 2 above)
Comment (bh2smith) [reply dune_client/base_client.py:17]: I like this, however, do you think we could make the log level configurable (and not hard coded as DEBUG).

File: dune_client/client_async.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
+""""
+Basic Dune Client Class responsible for refreshing Dune Queries
+Framework built on Dune's API Documentation
+https://duneanalytics.notion.site/API-Documentation-1b93d16e0fa941398e15047f643e003a
+"""
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply dune_client/client_async.py:6]: Yea... I am not sure why, but my IDE (PyCharm) still gives me a lot of squiggly red lines when I don't include this (even with >= 3.10). Any tips would be warmly welcomed.

File: dune_client/interface.py (hunk 3 above)
Comment (bh2smith) [reply dune_client/interface.py:24]: > same as sync interface but has async interface for refresh?
is this a question?

Yea, I think it would make plenty of sense to add more functions here. I am not exactly sure why I haven't. Did you have something particular in mind (apart from the obvious execute, get_status, cancel and get_results - which I agree are totally missing from here)? We could even remove that pylint disable!

**Review by bh2smith — APPROVED** [review]
I think this is coming along very well! Nice work here and nice code style. Your improvement suggestions are fantastic. Given that it is passing all the tests, I would say its ready for merge. Are there any e2e tests you think we might be missing?

Also, please take a look at the [CLA thing](https://github.com/cowprotocol/dune-client/pull/31#issuecomment-1301856544)

File: dune_client/base_client.py (hunk 2 above)
Comment (eliseygusev) [reply dune_client/base_client.py:17]: we could, also we can leave it as default and put logger.info/logger.debug where its suitable

File: dune_client/client_async.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eliseygusev) [reply dune_client/client_async.py:6]: interesting, I thought this __future__ import was only relevant for older python versions, let me check

**Comment by bh2smith** [comment]
It looks like the e2e tests fail because your PR is not allow to make use of project secrets. I will run this locally and check it out.

File: dune_client/interface.py (hunk 3 above)
Comment (eliseygusev) [reply dune_client/interface.py:24]: To be hounest I think I lack on the knowledge why is there an interface in the first place. So decided to let it be first, may be we can completely drop the interface and move all the logic to base class?

File: .gitignore [hunk 5]
```diff
@@ -1,4 +1,6 @@
 .env
 __pycache__/
 dist
-*.egg-info
\ No newline at end of file
+*.egg-info
+_version.py
+.idea/
```
[the remark this answers is not in the document]
Comment (eliseygusev) [reply .gitignore:6]: Yeah, thanks for the advice! Didn't know one could do that! Will add those for sure

File: .pylintrc (hunk 1 above)
Comment (bh2smith) [reply .pylintrc:2]: Do you think we could put it in the files containing the warning directly? This does seem a too much of a blanket solution.

File: dune_client/interface.py (hunk 3 above)
Comment (bh2smith) [reply dune_client/interface.py:24]: Maybe we can keep the original (non async interface around because https://github.com/cowprotocol/dune-alerts depends on it (and it would be a breaking change), but perhaps there is no reason to introduce the async interface from the get go, and we can phase it out in a major version bump later.

**Comment by bh2smith** [comment]
It would also be nice to see a usage test: 

Like 
```
async_dune.refresh(query1, query2, query3)  <- Did you provide an interface to query many?
```

and also some test that demonstrates the maximum number of parallel execution limitations.

**Comment by eliseygusev** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by eliseygusev** [comment]
recheck

**Comment by eliseygusev** [comment]
Fixed the issue with the tests failing -- forgot that ClientResponse().json() outputs the coroutine itself. I still have issues testing this locally as I don't have the Dune API key (and this key can't be easily acquired).

Didn't add any new functionality so far. I think it's better to leave it for future updates (as this PR already is kinda big).

**Comment by eliseygusev** [comment]
>async_dune.refresh(query1, query2, query3)  <- Did you provide an interface to query many?

as for this issue. I think the more convenient way would be for the consumer to do something like
`asyncio.gather(*[async_dune.refresh(query) for query in queries])`
This way the consumer has all the control around executing and limitations

File: dune_client/base_client.py (hunk 2 above)
Comment (bh2smith) [reply dune_client/base_client.py:14]: Ok, so I just realized that you actually removed the hard-coded DEBUG log level:

```
log = logging.getLogger(__name__)
logging.basicConfig(
    format="%(asctime)s %(levelname)s %(name)s %(message)s", level=logging.DEBUG
)
```
So, this is fine as is! Sorry for the confusion here.

File: dune_client/base_client.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        """Return default headers containing Dune Api token"""
+        return {"x-dune-api-key": self.token}
+
+    def query_to_params(self, query: Query) -> Dict[str, dict]:
+        """Transforms Query objects to params to pass in API"""
+        return {
+            "query_parameters": {
+                p.key: p.to_dict()["value"] for p in query.parameters()
+            }
+        }
```
Comment (bh2smith) [inline dune_client/base_client.py:39]: This method belongs on the query object (note that it does not make use of self) and has a single "query" attribute.

Comment (eliseygusev) [reply dune_client/base_client.py:39]: Yeah, that's true! Thanks for noticing. I guess we can move it to be a static method of Query dataclass?

File: dune_client/client_async.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+from dune_client.query import Query
+
+
+# pylint: disable=duplicate-code
+class AsyncDuneClient(BaseDuneClient):
+    """
+    An asynchronous interface for Dune API with a few convenience methods
+    combining the use of endpoints (e.g. refresh)
+    """
+    _session = None
```
Comment (bh2smith) [inline dune_client/client_async.py:29]: Putting the session here as None turns the type into an `Optional[ClientSession]` and causes a lot of hassle when trying to access session throughout the code. Is there any particular reason it needs to be defined out here initially as None? I believe all these issues would go away if it were merely defined in the constructor.

File: dune_client/client_async.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    async def close_session(self):
+        """Closes client session"""
+        await self._session.close()
+
+    async def __aenter__(self):
+        self._session = self._create_session()
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.close_session()
```
Comment (bh2smith) [inline dune_client/client_async.py:62]: Could you elaborate a bit on the meaning of these?

**Comment by bh2smith** [comment]
Here is the failing assertion error details:
```py
            assert (
                self._base_url.origin() == self._base_url
            ), "Only absolute URLs without path part are supported"
```
and the values are:
```
> self._base_url 
https://api.dune.com/api/v1 
> self._base_url.origin()
https://api.dune.com
```

I have also fixed all the lint, format and type errors on my side. Not sure if I can push them here (or if you even want me to)

File: dune_client/client_async.py (hunk 7 above)
Comment (eliseygusev) [reply dune_client/client_async.py:29]: Yeah, that makes sense. Let me think if we can remove None here

File: dune_client/client_async.py (hunk 8 above)
Comment (eliseygusev) [reply dune_client/client_async.py:62]: Sure! It's for `async with` syntax. see https://peps.python.org/pep-0492/#asynchronous-context-managers-and-async-with.

**Comment by eliseygusev** [comment]
>I have also fixed all the lint, format and type errors on my side. Not sure if I can push them here (or if you even want me to)

Thanks a lot for helping! Fixing the mistakes. Sorry that you have to manually run the pipeline, I just can not test this locally, so such mistakes occur... I would push code formatting together with tests fixes

File: dune_client/client_async.py (hunk 8 above)
Comment (bh2smith) [reply dune_client/client_async.py:62]: Nice thanks! I guess this is analogous to `with open(file...):` and does session closing instead of file closing etc... Makes perfect sense.

**Comment by bh2smith** [comment]
I think I could separate the pipeline a bit (partition lint format type instead of having them run sequentially so you can at least see all of the issues from a single push). Note that you should be able to run all but e2e checks on from your own side (i.e. without an API_KEY)

File: dune_client/base_client.py (hunk 6 above)
Comment (bh2smith) [reply dune_client/base_client.py:39]: Its not a static method of the query class (it takes a query as input -- i.e. self)

**Comment by bh2smith** [comment]
Hows this going @eliseygusev - Have you seen the PR with Fixes to make this pass CI?

https://github.com/eliseygusev/dune-client/pull/1


Would love to get this merged!

**Comment by eliseygusev** [comment]
Yeah, I’ve seen the PR. Would update the code either today in the evening
or tomorrow. Thanks for your support!

**Comment by eliseygusev** [comment]
Ok, so merged your request. Thanks a lot for it! But i'm kind of lost about whether there is something left to fix before merging?

File: dune_client/base_client.py (hunk 6 above)
Comment (eliseygusev) [reply dune_client/base_client.py]: Resolved this and added unit-test for the functionality

**Comment by eliseygusev** [comment]
Fixed the linters but not sure whats happening with the tests. Is it permissions' things or something is failing in the code?

**Comment by bh2smith** [comment]
I think its all good. We can merge and publish a beta release from there. I won't have time to test it out this week though, but at least it will be available.

**Comment by bh2smith** [comment]
It is a permission thing, your branch does not have access to our secrets.

**Comment by bh2smith** [comment]
but you can also run these lint checks on your own side before pushing:

```
black ./ && pylint dune_client && mypy dune_client --strict
```

**Comment by eliseygusev** [comment]
>but you can also run these lint checks on your own side before pushing:

black ./ && pylint dune_client && mypy dune_client --strict

yeah, sorry, my bad. always forget about it. can I add Makefile with commands for linters and tests so that one can just
`make lint` to make the code pass linters?

**Comment by bh2smith** [comment]
> yeah, sorry, my bad. always forget about it. can I add Makefile with commands for linters and tests so that one can just
make lint to make the code pass linters?

That's great, but in a separate PR.

**Comment by eliseygusev** [comment]
>That's great, but in a separate PR.

Ok, will add it soon. Is there anything else in this PR left to do on my side? Or you will just merge and publish by yourself?

**Comment by bh2smith** [comment]
> Is there anything else in this PR left to do on my side?

Fix the failing lint! I will do the rest.

**Comment by eliseygusev** [comment]
>Fix the failing lint! I will do the rest.

I wasn't sure why mypy failed in CI. I didnt touch interface.py and locally mypy --strict works. However, I think I did fix it with the addition of `abc.abstract_method` should be good now

**Comment by bh2smith** [comment]
Has been published under https://pypi.org/project/dune-client/0.2.0b0/

Thanks very much for your contribution!

**Comment by bh2smith** [comment]
I am trying to use this async client and keep running into these errors

```

RuntimeError: Timeout context manager should be used inside a task
2022-11-16 00:49:21,368 ERROR asyncio Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x106ba64a0>
```

I have tried putting a close session after the refresh call, like so

```
response = await self.dune.refresh(query, ping_frequency=10)
await self.dune.close_session()
```

but still get them. Would love if you could show me how to use this from an external package and maybe make it somehow more convenient for external users.
